### PR TITLE
NOCI python bindings

### LINKED ISSUE #1008: Implement python bindings for the NOCI method
After all necessary components are implemented, we can add the python bindings as well.

### PR DESCRIPTION
**Pull request by xdvriend**
**Short description**

With the NOCI method now fully implemented in the C++ library, this PR aims to bind that functionality to python.

**Related issues**

- closes #1008 

**To do**

- [x] (if applicable, when creating new source files): don't forget to update the Doxygen file, the collective `gqcp.hpp` include header and the CMake files
- [x] implement bindings for `NonOrthogonalStateBases`
- [x] implement bindings for the NOCI method
- [x] implement bindings for the NOCIExpansion
- [x] Provide a NOCI example notebook one the website

### CODE REVIEW

File: gqcpy/include/interfaces.hpp [hunk 1]
```diff
@@ -1129,4 +1129,85 @@ void bindQCModelHartreeFockRealStabilityInterface(Class& py_class) {
...
+             &Type::basisStateDimension,
+             "Return the dimension of the basis states in this non-orthogonal basis.")
+
+        .def("metric",
+             &Type::metric,
+             "Return the overlap metric of this non-orthogonal basis as an operator.")
+
+        .def("numberOfBasisStates",
+             &Type::numberOfBasisStates,
+             "Return the number of basis staes in this non-orthogonal basis.")
```
Comment (johdvos) [inline gqcpy/include/interfaces.hpp]:
```suggestion
             "Return the number of basis states in this non-orthogonal basis.")
```

File: website/examples/NOCI.ipynb [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    "spinor_basis = gqcpy.GSpinorBasis_d(molecule, \"6-31G\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this general spinor basis, we can do two more things we will need for our NOCI calculation:\n",
+    "- Quantize the overlap operator, which will serve as a metric in the NOCI basis\n",
+    "- Quantize th Hamiltonian, which we will need for the eventual calculation"
```
Comment (johdvos) [inline website/examples/NOCI.ipynb]:
```suggestion
    "- Quantize the Hamiltonian, which we will need for the eventual calculation"
```

**Review by johdvos — APPROVED** [review]
Everything seems alright to me (with limited knowledge of C++ of course). Fixed two documentation typos.

**Comment by xdvriend** [comment]
> Everything seems alright to me (with limited knowledge of C++ of course). Fixed two documentation typos.

Thank you for reviewing this! I hope this gave you some insight in the inner NOCI workings of GQCP.

**Pull request merged**